File: clients/android/app/src/main/java/com/remoteeverything/app/CatalogController.kt
package com.remoteeverything.app

import com.remoteeverything.core.api.GatewayClient
import com.remoteeverything.core.api.CatalogOutcome
import com.remoteeverything.core.api.ControlCode
import com.remoteeverything.core.api.toAppInfo
import com.remoteeverything.core.model.AppInfo
import com.remoteeverything.core.model.AppState
import com.remoteeverything.core.model.Cadence
import com.remoteeverything.core.model.ClientError
import com.remoteeverything.core.model.Identity
import com.remoteeverything.core.model.Node
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Manages the live catalog of one node: periodic background refresh, application
 * start/stop control, and settle polling until the app reaches a steady state.
 */
class CatalogController(
    private val scope: CoroutineScope,
    private val nodesController: NodesController,
    private val currentNetworkKey: () -> String,
    private val currentIdentities: suspend () -> List<Identity>,
    private val currentNode: (String) -> Node?,
    private val onNotice: (Int) -> Unit,
) {
    private val _catalog = MutableStateFlow<CatalogUiState>(CatalogUiState.Loading)
    val catalog: StateFlow<CatalogUiState> = _catalog.asStateFlow()

    private var catalogJob: Job? = null
    private var watchingNodeId: String? = null

    /**
     * Watches one node's catalog.
     *
     * The node is read again on every tick, and that is not a detail: a catalog is
     * fetched *through* one of the node's paths, and which paths are reachable is a
     * fact about this minute. A screen that held on to the node object it was
     * opened with would be a screen that stays offline for as long as it is open —
     * without asking anything, because a path it believes dead is a request it
     * never makes — while the list behind it refreshes every minute and says
     * otherwise.
     *
     * [force] is what a person's Retry means: ask again, now. Without it this is
     * the idempotent version a screen asks for on entry, and a second call while
     * the loop is already running is nothing to do — which is exactly why Retry
     * has to say what it wants, or the button is a button that does not exist.
     */
    fun watchNode(node: Node, force: Boolean = false) {
        if (!force && watchingNodeId == node.id && catalogJob?.isActive == true) return
        val nodeId = node.id
        watchingNodeId = nodeId
        catalogJob?.cancel()
        _catalog.value = CatalogUiState.Loading
        catalogJob = scope.launch {
            while (true) {
                _catalog.value = fetchCatalog(currentNode(nodeId) ?: node)
                delay(Cadence.catalogRefreshMs)
            }
        }
    }

    fun stopWatching() {
        catalogJob?.cancel()
        catalogJob = null
        watchingNodeId = null
    }

    /**
     * One node's applications, asked for down every road that answers.
     *
     * The chosen path goes first, and a path that stopped answering is a reason to
     * try the one behind it — a phone that still holds a tunnel must not be told
     * its machine is gone because the Wi-Fi it was on is gone. A refusal is
     * different: it is the gateway's answer, and another road would answer the same.
     */
    suspend fun fetchCatalog(node: Node): CatalogUiState {
        val best = nodesController.choosePath(node, currentNetworkKey())
        val roads = (listOfNotNull(best) + node.paths.filter { it.reachable == true && it.origin != best?.origin })
        if (roads.isEmpty()) return CatalogUiState.Offline
        var answer: CatalogUiState = CatalogUiState.Offline
        for (path in roads) {
            val identity = currentIdentities().firstOrNull { it.origin == path.origin } ?: continue
            val client = nodesController.clientFor(identity) ?: continue
            try {
                return CatalogOutcome.forAnswer(client.catalog(node.id)).toUiState()
            } catch (e: ClientError) {
                answer = CatalogOutcome.forRefusal(e).toUiState()
                // An answer that is not "the machine is off" is the gateway talking:
                // asking it again by another road would say the same thing.
                if (answer !is CatalogUiState.Offline) return answer
            } catch (e: Exception) {
                answer = CatalogUiState.Offline
            }
            // This road is not answering: the next one gets the turn, and the
            // remembered choice stops pointing here.
            nodesController.dropChoice(node.id)
        }
        return answer
    }

    /**
     * Asks the node being watched again, now: the phone's network changed, so the
     * road in use is a question rather than an answer.
     */
    fun recheckNow() {
        val nodeId = watchingNodeId ?: return
        val node = currentNode(nodeId) ?: return
        watchNode(node, force = true)
    }

    /**
     * Starting and stopping are asked for, then waited on. The row says what was
     * asked for at once — `starting` or `stopping` — because the alternative is a
     * button that keeps offering a start that has already been sent, which is both
     * a puzzle and a way to send it twice. What follows is the node's own answer:
     * every poll writes the state the node reports, and the catalog is read again
     * when the application settles.
     */
    fun control(node: Node, appId: String, start: Boolean) {
        applyState(appId, if (start) AppState.STARTING else AppState.STOPPING)
        scope.launch {
            val path = nodesController.choosePath(node, currentNetworkKey())
            if (path == null) {
                _catalog.value = CatalogUiState.Offline
                return@launch
            }
            val identity = currentIdentities().firstOrNull { it.origin == path.origin }
            val client = identity?.let { nodesController.clientFor(it) }
            if (client == null) {
                _catalog.value = fetchCatalog(node)
                return@launch
            }
            val answer = try {
                if (start) client.start(node.id, appId) else client.stop(node.id, appId)
            } catch (e: Exception) {
                onNotice(R.string.error_control_failed)
                _catalog.value = fetchCatalog(node)
                return@launch
            }
            if (!answer.ok && answer.code == ControlCode.STATE_UPDATE_FAILED) {
                onNotice(R.string.error_stop_failed)
            }
            if (answer.code == ControlCode.COMPUTER_OFFLINE) {
                _catalog.value = CatalogUiState.Offline
                return@launch
            }
            if (answer.code == ControlCode.APP_NOT_FOUND) {
                onNotice(R.string.error_app_gone)
                _catalog.value = fetchCatalog(node)
                return@launch
            }
            // Starting and stopping are not instant: the screen keeps asking until
            // the application settles, so the row never shows a state it left.
            pollUntilSettled(client, node.id, appId)
            _catalog.value = fetchCatalog(currentNode(node.id) ?: node)
        }
    }

    private suspend fun pollUntilSettled(
        client: GatewayClient,
        nodeId: String,
        appId: String,
    ) {
        var waited = 0L
        var interval = Cadence.controlPollMs
        while (waited < Cadence.controlPollTimeoutMs) {
            delay(interval)
            waited += interval
            val status = try {
                client.status(nodeId, appId)
            } catch (e: Exception) {
                return
            }
            if (status.code == ControlCode.COMPUTER_OFFLINE) {
                _catalog.value = CatalogUiState.Offline
                return
            }
            if (status.code == ControlCode.APP_NOT_FOUND) {
                onNotice(R.string.error_app_gone)
                return
            }
            // The node's answer is the row's state, not the client's guess.
            status.app?.let { applyState(appId, it.toAppInfo().code) }
            if (status.code == ControlCode.READY || status.code == ControlCode.STOPPED) return
            interval = (interval * Cadence.controlPollFactor).toLong().coerceAtMost(Cadence.controlPollCeilingMs)
        }
    }

    /** Writes one application's state into the catalog that is on screen. */
    private fun applyState(appId: String, code: AppState) {
        val current = _catalog.value
        if (current !is CatalogUiState.Ready) return
        _catalog.value = CatalogUiState.Ready(
            current.apps.map { app -> if (app.id == appId) app.copy(code = code) else app },
        )
    }
}

private fun CatalogOutcome.toUiState(): CatalogUiState = when (this) {
    is CatalogOutcome.Apps -> CatalogUiState.Ready(apps)
    CatalogOutcome.Offline -> CatalogUiState.Offline
    CatalogOutcome.Unauthorized -> CatalogUiState.Unauthorized
    CatalogOutcome.GatewayTrouble -> CatalogUiState.Failed(null)
}

File: clients/ios/RemoteEverything/App/CatalogController.swift
import Foundation

@MainActor
final class CatalogController: ObservableObject {

    @Published private(set) var catalogs: [String: CatalogState] = [:]

    private let nodesController: NodesController
    private let clients: GatewayClientPool
    private let vault: IdentityVault
    private let pathSelector: PathSelector.Cache

    var onNotice: ((Notice) -> Void)?

    private var catalogLoop: Task<Void, Never>?
    private var controlTasks: [String: Task<Void, Never>] = [:]
    private var catalogNodeID: String?

    init(
        nodesController: NodesController,
        clients: GatewayClientPool,
        vault: IdentityVault,
        pathSelector: PathSelector.Cache
    ) {
        self.nodesController = nodesController
        self.clients = clients
        self.vault = vault
        self.pathSelector = pathSelector
    }

    func stopLoops() {
        catalogLoop?.cancel()
        catalogLoop = nil
        for task in controlTasks.values { task.cancel() }
        controlTasks.removeAll()
    }

    func clear() {
        catalogs.removeAll()
    }

    func setCatalog(_ state: CatalogState, forNodeID nodeID: String) {
        catalogs[nodeID] = state
    }

    // MARK: - Polling & Catalog Loading

    func beginCatalogPolling(nodeID: String) {
        catalogNodeID = nodeID
        catalogLoop?.cancel()
        catalogLoop = Task { [weak self] in
            while !Task.isCancelled {
                guard let self else { return }
                await self.loadCatalog(nodeID: nodeID)
                do {
                    try await Task.sleep(nanoseconds: Cadence.catalogRefreshNanos)
                } catch {
                    return
                }
            }
        }
    }

    func endCatalogPolling() {
        catalogLoop?.cancel()
        catalogLoop = nil
        catalogNodeID = nil
    }

    /// Asks the node being watched again, now: the phone's network changed, so
    /// the road in use is a question rather than an answer (Android recheckNow).
    func recheckNow() {
        guard let nodeID = catalogNodeID else { return }
        Task { await loadCatalog(nodeID: nodeID) }
    }

    func loadCatalog(nodeID: String) async {
        guard let node = nodesController.node(withID: nodeID) else {
            catalogs[nodeID] = .offline
            return
        }
        if catalogs[nodeID] == nil { catalogs[nodeID] = .loading }

        // The chosen path goes first, then the other roads that answered; a
        // gateway that says this device may not reach the node says so once,
        // and the next road does not get asked the same question twice.
        var lastState: CatalogState = .offline
        for path in nodesController.orderedPaths(for: node) {
            guard let identity = nodesController.identity(forPath: path),
                  let client = clients.deviceClient(for: identity, vault: vault)
            else {
                nodesController.setCondition(origin: path.origin, condition: .needsPairing)
                continue
            }
            let started = Date()
            do {
                let response = try await client.catalog(nodeID: nodeID)
                nodesController.markReachable(path: path, nodeID: nodeID, latencyMs: Int(Date().timeIntervalSince(started) * 1000))
                nodesController.setCondition(origin: identity.origin, condition: .ok)
                catalogs[nodeID] = CatalogController.state(for: response)
                return
            } catch let error as ClientError {
                let state = CatalogController.state(for: error)
                // A refusal that is not "the machine is off" is the gateway
                // talking: asking it again by another road would say the same.
                if state != .offline {
                    if case .unauthorized = state {
                        nodesController.setCondition(origin: identity.origin, condition: .unauthorized)
                    }
                    catalogs[nodeID] = state
                    return
                }
                lastState = state
            } catch {
                // Could not be reached at all — the network, not the node: read
                // as an offline machine, the way the other two clients do.
                lastState = .offline
            }
            nodesController.markUnreachable(path: path, nodeID: nodeID)
            pathSelector.invalidate(nodeID: nodeID)
        }
        catalogs[nodeID] = lastState
    }

    // MARK: - App Control

    func startApplication(nodeID: String, appID: String) {
        control(nodeID: nodeID, appID: appID, action: .start)
    }

    func stopApplication(nodeID: String, appID: String) {
        control(nodeID: nodeID, appID: appID, action: .stop)
    }

    private enum ControlAction {
        case start
        case stop

        var pendingState: AppState { self == .start ? .starting : .stopping }
    }

    private func control(nodeID: String, appID: String, action: ControlAction) {
        let key = "\(nodeID)/\(appID)"
        controlTasks[key]?.cancel()
        updateApplication(nodeID: nodeID, appID: appID) { app in
            var updated = app
            updated.code = action.pendingState
            updated.enabled = app.enabled
            updated.running = app.running
            return updated
        }
        controlTasks[key] = Task { [weak self] in
            guard let self else { return }
            await self.performControl(nodeID: nodeID, appID: appID, action: action)
        }
    }

    private func performControl(nodeID: String, appID: String, action: ControlAction) async {
        guard let node = nodesController.node(withID: nodeID) else { return }
        // A start or a stop goes down the one road the phone would take; asking
        // again by another road is what the catalog refresh does next.
        guard let path = nodesController.preferredPath(for: node),
              let identity = nodesController.identity(forPath: path),
              let client = clients.deviceClient(for: identity, vault: vault)
        else {
            await loadCatalog(nodeID: nodeID)
            return
        }
        do {
            let response: ControlResponse
            switch action {
            case .start: response = try await client.start(nodeID: nodeID, appID: appID)
            case .stop: response = try await client.stop(nodeID: nodeID, appID: appID)
            }
            if response.code == .computerOffline {
                catalogs[nodeID] = .offline
                return
            }
            if response.code == .appNotFound {
                onNotice?(.appGone)
                await loadCatalog(nodeID: nodeID)
                return
            }
            // The gateway says the state would not change — a stop that cannot
            // stop, a start that cannot start — the same two spellings the wire
            // has for it, caught the way Android catches the first.
            if response.code == .stateUpdateFailed || response.errorCode == "stop_command_failed" {
                onNotice?(.stopFailed)
            }
            if let app = response.app {
                applyApplication(nodeID: nodeID, info: app.appInfo)
            } else {
                updateApplication(nodeID: nodeID, appID: appID) { app in
                    var updated = app
                    updated.enabled = response.enabled
                    updated.running = response.running
                    updated.code = state(enabled: response.enabled, running: response.running)
                    return updated
                }
            }
            await pollStatus(nodeID: nodeID, appID: appID, client: client, path: path)
            return
        } catch {
            // A start or a stop that did not go through is one sentence, whatever
            // the road it failed on: the catalog is read again, which is what
            // says whether the node went offline or the device was refused.
            onNotice?(.controlFailed)
        }
        await loadCatalog(nodeID: nodeID)
    }

    private func pollStatus(nodeID: String, appID: String, client: GatewayClient, path: Path) async {
        let deadline = Date().addingTimeInterval(Cadence.controlPollTimeout)
        var interval = Cadence.controlPoll
        while !Task.isCancelled, Date() < deadline {
            do {
                try await Task.sleep(nanoseconds: Cadence.nanoseconds(seconds: interval))
            } catch {
                return
            }
            do {
                let response = try await client.status(nodeID: nodeID, appID: appID)
                if response.code == .computerOffline {
                    catalogs[nodeID] = .offline
                    return
                }
                if response.code == .appNotFound {
                    onNotice?(.appGone)
                    await loadCatalog(nodeID: nodeID)
                    return
                }
                if let app = response.app {
                    applyApplication(nodeID: nodeID, info: app.appInfo)
                    if app.code.isSteady { return }
                } else {
                    updateApplication(nodeID: nodeID, appID: appID) { app in
                        var updated = app
                        updated.enabled = response.enabled
                        updated.running = response.running
                        updated.code = state(enabled: response.enabled, running: response.running)
                        return updated
                    }
                    if state(enabled: response.enabled, running: response.running).isSteady { return }
                }
            } catch let error as ClientError {
                if error.code == .unauthorized {
                    catalogs[nodeID] = .unauthorized
                    nodesController.setCondition(origin: path.origin, condition: .unauthorized)
                    return
                }
                if error.code == .rateLimited || error.code == .serverBusy {
                    onNotice?(.busy)
                }
            } catch {
            }
            interval = min(interval * Cadence.controlPollFactor, Cadence.controlPollCeiling)
        }
    }

    private func state(enabled: Bool, running: Bool) -> AppState {
        switch (enabled, running) {
        case (true, true): return .ready
        case (true, false): return .starting
        case (false, true): return .stopping
        case (false, false): return .stopped
        }
    }

    private func applyApplication(nodeID: String, info: AppInfo) {
        guard case .ready(var apps) = catalogs[nodeID] else { return }
        if let index = apps.firstIndex(where: { $0.id == info.id }) {
            apps[index] = info
            catalogs[nodeID] = .ready(apps)
        }
    }

    private func updateApplication(nodeID: String, appID: String, transform: (AppInfo) -> AppInfo) {
        guard case .ready(var apps) = catalogs[nodeID],
              let index = apps.firstIndex(where: { $0.id == appID })
        else { return }
        apps[index] = transform(apps[index])
        catalogs[nodeID] = .ready(apps)
    }

    private static func state(for response: CatalogResponse) -> CatalogState {
        switch CatalogOutcome.forAnswer(response) {
        case .apps(let apps): return .ready(apps)
        case .offline: return .offline
        case .unauthorized: return .unauthorized
        case .gatewayTrouble: return .gatewayTrouble
        }
    }

    private static func state(for error: ClientError) -> CatalogState {
        switch CatalogOutcome.forRefusal(error) {
        case .offline: return .offline
        case .unauthorized: return .unauthorized
        case .apps, .gatewayTrouble: return .gatewayTrouble
        }
    }

    // MARK: - Web Target Resolution

    /// Where one application lives, asked down the one road the phone would take
    /// — the same single road Android's `open` uses. A road that stops answering
    /// is the catalog refresh's question, not this one's.
    func resolveWebTarget(nodeID: String, appID: String) async throws -> WebTarget {
        guard let node = nodesController.node(withID: nodeID) else { throw ClientError(.nodeNotFound) }
        guard let path = nodesController.preferredPath(for: node),
              let identity = nodesController.identity(forPath: path),
              let credential = vault.prepare(origin: identity.origin),
              let client = clients.deviceClient(for: identity, vault: vault)
        else {
            throw ClientError(.unauthorized)
        }
        let url = try await client.open(nodeID: nodeID, appID: appID)
        var target = url
        if case .ready(let apps) = catalogs[nodeID],
           let app = apps.first(where: { $0.id == appID }),
           !app.launchFragment.isEmpty,
           let withFragment = URL(string: url.absoluteString + app.launchFragment) {
            target = withFragment
        }
        let handler = GatewayChallengeHandler(
            gatewayOrigin: identity.origin,
            pin: identity.serverPin,
            credential: credential
        )
        return WebTarget(gatewayOrigin: identity.origin, url: target, handler: handler, isPrivate: path.isPrivate)
    }
}
